synapse/rest/client/v2_alpha/report_event.py

```diff
@@ -58,17 +81,108 @@ async def on_POST(self, request, room_id, event_id):
                 Codes.BAD_JSON,
             )
 
+        target = body.get("org.matrix.msc2938.target", "homeserver-admins")
+
+        if target == "room-moderators":
+            # Report event to room moderators.
+            # This branch has further safety checks (e.g. can the user actually see the event?)
+            return await self._target_room_moderators(room_id=room_id, event_id=event_id, user_id=user_id, reason=body["reason"], score=body["score"])
+        elif target == "homeserver-admins":
+            return await self._target_homeserver_admin(room_id=room_id, event_id=event_id, user_id=user_id, reason=body["reason"], content=body)
+        else:
+            raise SynapseError(
+                HTTPStatus.BAD_REQUEST,
+                "Optional param 'target' must be one of ['homeserver-admins', 'room-moderators']",
+                Codes.BAD_JSON,
+            )
+
+    async def _target_homeserver_admin(self, room_id, event_id, user_id, reason, content):
+        # Store event report for later investigation by the homeserver admin.
         await self.store.add_event_report(
             room_id=room_id,
             event_id=event_id,
             user_id=user_id,
-            reason=body["reason"],
-            content=body,
+            reason=reason,
+            content=content,
             received_ts=self.clock.time_msec(),
         )
 
         return 200, {}
 
+    async def _target_room_moderators(self, room_id, event_id, user_id, reason, score):
+        # We're dispatching the abuse report to room moderators.
+        # First, let's make sure that we should.
+
+        # A little sanity check on the event itself.
+        event = await self.store.get_event(event_id)
+        if event.room_id != room_id:
+            raise SynapseError(
+                HTTPStatus.BAD_REQUEST,
+                "No such event in this room",
+                Codes.NOT_FOUND
+            )
+
+        # Now make sure that the user was able to witness the event.
+        events = await filter_events_for_client(
+            self.storage,
+            user_id,
+            [event],
+            is_peeking=True
+        )
+        if len(events) == 0:
+            raise SynapseError(
+                HTTPStatus.FORBIDDEN,
+                "User cannot witness this event",
+                Codes.FORBIDDEN
+            )
+
+        # Find the moderators
+        power_level_event = await self.state.get_current_state(
+            room_id, EventTypes.PowerLevels, ""
+        )
+        ban_level = power_level_event.content.get("ban", 50)
+        kick_level = power_level_event.content.get("kick", 50)
+        moderator_level = max(ban_level, kick_level)
```

Comment:
Does this definition of moderator match other places in the matrix spec?

Reply:
As far as I can tell, there is no definition of moderator in other places of the matrix spec:

> Clients may wish to assign names to particular power levels. A suggested mapping is as follows: - 0 User - 50 Moderator - 100 Admin

Is pretty much it. Part of the MSC actually defines what a moderator is: someone who can both kick and ban.